crates/astria-sequencer-relayer/src/relayer/celestia_client/mod.rs

```diff
@@ -0,0 +1,654 @@
+    ///
+    /// The `last_error_receiver` will provide the error from the previous attempt if this is not
+    /// the first attempt for these blobs, or `None` if it is the first attempt.  The error can be
+    /// used to obtain the appropriate fee in the case that the previous attempt failed due to a
+    /// low fee.
+    // Copied from https://github.com/celestiaorg/celestia-app/blob/v1.4.0/x/blob/payforblob.go
+    pub(super) async fn try_submit(
+        mut self,
+        blobs: Arc<Vec<Blob>>,
+        state: Arc<super::State>,
+        last_error_receiver: watch::Receiver<Option<Error>>,
+    ) -> Result<u64, Error> {
+        // Get the error from the last attempt to `try_submit` and if `Some`, just update the
+        // cached cost params.
+        let maybe_last_error = last_error_receiver.borrow().clone();
+        if maybe_last_error.is_some() {
+            self.fetch_and_cache_cost_params(state.clone()).await?;
+        }
+
+        let msg_pay_for_blobs = new_msg_pay_for_blobs(blobs.iter(), self.address.clone())?;
+
+        let base_account = self.fetch_account().await?;
+        debug!(
+            account_number = base_account.account_number,
+            sequence = base_account.sequence,
+            "got account info"
+        );
+
+        let cost_params = state.celestia_cost_params();
+        let gas_limit = estimate_gas(&msg_pay_for_blobs.blob_sizes, cost_params);
+        let fee = calculate_fee(cost_params, gas_limit, maybe_last_error);
+        trace!(?cost_params, gas_limit = gas_limit.0, fee_utia = fee);
+        debug!(fee_utia = fee, "estimated fee");
+
+        let signed_tx = new_signed_tx(
+            &msg_pay_for_blobs,
+            &base_account,
+            gas_limit,
+            fee,
+            self.chain_id.clone(),
+            &self.signing_keys,
+        );
+
+        let blob_tx = new_blob_tx(&signed_tx, blobs.iter());
+
+        let tx_hash = self.broadcast_tx(blob_tx).await?;
+
+        loop {
+            tokio::time::sleep(Duration::from_secs(1)).await;
+            if let Some(height) = self.get_tx(tx_hash.clone()).await? {
+                // allow: the height is never negative.
+                #[allow(clippy::cast_sign_loss)]
+                return Ok(height as u64);
+            }
+        }
+    }
+
+    /// Makes queries to the Celestia app to retrieve pricing information, which is cached in
+    /// `state` on success.
+    ///
+    /// This should ideally only be called once during the session (at startup by
+    /// `CelestiaClientBuilder::try_build`), but will also be called if an attempt to submit blobs
+    /// fails.
+    async fn fetch_and_cache_cost_params(&mut self, state: Arc<super::State>) -> Result<(), Error> {
+        let gas_per_blob_byte = self.fetch_blob_params().await?.gas_per_blob_byte;
+        let tx_size_cost_per_byte = self.fetch_auth_params().await?.tx_size_cost_per_byte;
+        let min_gas_price = self.fetch_min_gas_price().await?;
```

Comment:
I think I'd have to clone the underlying gRPC clients to avoid multiple mutable borrows of self, but that should be fairly cheap, so probably worthwhile.

Reply:
Done in [60359ea](https://github.com/astriaorg/astria/pull/963/commits/60359ea9c9c69bda0359b474d55eccacb77514e7).